### PR TITLE
ui: Add a tasks page

### PR DESCRIPTION
**Pull request by dmsimard**
This adds a new task "index" page much like the one for playbooks and hosts. It is modelled after ``/api/v1/tasks``.

It's up for testing purposes on [demo.recordsansible.org/tasks](https://demo.recordsansible.org/tasks) right now and looks like this:
Peek 2021-12-04 19-56

### CODE REVIEW

File: ara/ui/templates/partials/task_status_icon.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<a
+  {% if not static_generation %}href="{% url 'ui:task_index' %}?status={{ status }}"{% endif %}
+  class="btn btn-sm {% if status == 'completed' %}btn-success{% elif status == 'failed' %}btn-danger{% elif status == 'running' %}btn-info{% else %}btn-warning{% endif %}"
```
Comment (dmsimard) [inline ara/ui/templates/partials/task_status_icon.html:3]: There's no failed status for a task:
```suggestion
  class="btn btn-sm {% if status == 'completed' %}btn-success{% elif status == 'running' %}btn-info{% else %}btn-warning{% endif %}"
```

**Comment by coxoperationsengineer** [comment]
Thanks for adding this @dmsimard this is a game changer for sifting through a complex play across a large number of hosts!

**Comment by hille721** [comment]
I like this as well! But would love to see a playbook name/ path column and a filter for playbook name/ path as well. 
You will have same tasks in different playbooks, so it will be really helpfull to directly see which playbook it is. 

Btw for the hosts index view I also miss that, as already mentioned in #331.

**Comment by dmsimard** [comment]
@hille721 yes, I think there is work to do to make the pages have a consistent feel in general.

What bothers me is that the status is moderately misleading because it can only be one of the following:
https://github.com/ansible-community/ara/blob/f366c9f9f2a06311dd67449f1124dca0674563c2/ara/api/models.py#L202

There is no concept of a failed task: only whether it's completed or not. The checkmark could be green even if there are failed results in the task. Not great.
Computing the status of a task server side would mean looking up each result of a task and return ``failed`` if there is a failure. Expensive and slow to do at scale.

The callback can determine the status of a task because there are specific hooks that are triggered on failure like: https://github.com/ansible-community/ara/blob/f366c9f9f2a06311dd67449f1124dca0674563c2/ara/plugins/callback/ara_default.py#L390
or:
https://github.com/ansible-community/ara/blob/f366c9f9f2a06311dd67449f1124dca0674563c2/ara/plugins/callback/ara_default.py#L399

This would mean adding a ``failed`` task status to the model and then implementing the logic in the callback to set the status of the task accordingly when updating it's status:
https://github.com/ansible-community/ara/blob/f366c9f9f2a06311dd67449f1124dca0674563c2/ara/plugins/callback/ara_default.py#L415-L429

I thought that this could be tricky when running with the free strategy and it reminded me that we haven't fixed ``end_task`` when running with the free strategy as per the discussions in https://github.com/ansible-community/ara/pull/268.

There is a bit of work to do :)

**Comment by dmsimard** [comment]
I forgot to say why I pulled this from 1.5.8: it sort of works but as mentioned in my last comment, there needs to be some changes under the hood to make it a better implementation.
The changes are significant enough to be in a larger version bump so we can ship some minor updates and bug fixes out sooner.

**Comment by netopsengineer** [comment]
> I forgot to say why I pulled this from 1.5.8: it sort of works but as mentioned in my last comment, there needs to be some changes under the hood to make it a better implementation.
> 
> The changes are significant enough to be in a larger version bump so we can ship some minor updates and bug fixes out sooner.

Thanks for your dedication to delivering quality features and updates @dmsimard

**Comment by dmsimard** [comment]
> There is no concept of a failed task: only whether it's completed or not. The checkmark could be green even if there are failed results in the task. Not great.

I realized we have the same issue with plays since there are only completed or running statuses. I'll focus on tasks first.

**Comment by dmsimard** [comment]
I've rebased the patch on top of 1.5.8rc1 and included a first iteration at the API and callback implementation for working with the new "failed" task status. It works though I will review and add at least some tests, most notably for the free strategy.

It looks like this:
Peek 2022-03-27 17-21

Note an interesting case in the demo -- there is a failed task but the playbook is successful. How come ?
When you look at the task file, it's because it is later rescued since the task is inside a block with a rescue and there's no ignore_errors.

The callback will only flip the status to "failed" if ignore_errors is unset or false.

**Comment by dmsimard** [comment]
I saw an opportunity to add a link to the new host and task search pages from the numbers on the playbook index so you can easily search for the hosts and tasks for a particular playbook:
Peek 2022-03-27 20-42

**Comment by dmsimard** [comment]
> It works though I will review and add at least some tests, most notably for the free strategy.

I was testing what this looked like with the free strategy and caught a somewhat expected bug where there is a failure in one of the results of the task but the task result was successful. Will try to see if I can fix it.

Screenshot from 2022-04-17 20-25-24

--

Screenshot from 2022-04-17 20-25-45

**Comment by dmsimard** [comment]
I've updated the database migration to include a script that will iterate over tasks and update the status to ``failed`` if there are any failed or unreachable results for that task.

I haven't yet found a solution to the free strategy issue that didn't involve making one or more API calls which I am trying to avoid in the best interest of performance. I am going to file it under known issues for now until we can spend more time fixing issues in the free strategy recording particularly.

I think this should be good to go as a first iteration. I will make a backup of the demo.recordsansible.org database and update it with the latest revision of this PR.

**Comment by dmsimard** [comment]
Enough iterating for today but I do want to point out before I forget: running the migrations updates the "updated" field for the tasks that are updated which... makes sense but can perhaps provide confusing results when looking at the tasks of a playbook:
Screenshot from 2022-04-17 21-55-16

**Comment by dmsimard** [comment]
While using the tasks page I realized I forgot to add a couple fields:

Screenshot from 2022-05-03 18-16-04

We should also be able to search by task file path and lineno, handler (boolean), as well as file (to view all the tasks from a file?) and play: https://github.com/ansible-community/ara/blob/db8243c3af938ece12c9cd59dd7fe4d9a711b76d/ara/api/models.py#L202-L213

We can't search by tags yet unfortunately. With the power of hindsight, the implementation shouldn't have been using compression (which was meant for larger fields like result content) and which makes it harder (and much more expensive) to search text.

I'm hoping to improve that eventually by relying on [django's 3.2's new json fields](https://github.com/ansible-community/ara/issues/153) but it will have to be in a "very major" release since it will require significant database and API changes.

**Comment by dmsimard** [comment]
Just rebased the PR because the branch started to be old/stale.

That issue about the last updated field being bumped during the migration kind of bothers me but at the same time, it's doing what is intended.

I guess we should change the field in the table to be the date of creation instead of the date of update ?

**Comment by dmsimard** [comment]
```
CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0011_play_label_name_max_length, 0011_task_status in api).
To fix them run 'python manage.py makemigrations --merge'
```

**Comment by dmsimard** [comment]
I've iterated on a batch of various changes, it probably needs some cleanup before landing but it's largely good to go for this first release.

This is what it looks like now (open gif in new tab if it's too small):
Peek 2022-09-10 16-06

**Comment by dmsimard** [comment]
I've deployed this PR to https://demo.recordsansible.org/tasks, can you let me know what you think ?

**Comment by dmsimard** [comment]
I've cleaned up the history, this is no longer WIP -- I've tested it and it is good to land IMO but I'll let it simmer a bit on the demo to see if folks have feedback.

**Comment by hille721** [comment]
Imo the playbook id filter can be removed. Not really usable by a human and we are having the playbook name filter anyhow

**Comment by dmsimard** [comment]
> Imo the playbook id filter can be removed. Not really usable by a human and we are having the playbook name filter anyhow

An id is not the same thing as a name, though: searching by id gets a single unique playbook whereas a name gets one or more playbooks. Right now the id is used to provide a direct link to tasks like this:

Peek 2022-09-11 15-52

We could consider making the field hidden ? This way we can filter by id (from other places) without needing a field for it.
But then, if the things are filtered by id, it is relevant to show somewhere that the output is filtered by something.

It's not really meant for humans, I guess, but I'm open to suggestions :p

File: ara/ui/templates/partials/task_index_row.html [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {% if not static_generation %}
+      <a href="{% url 'ui:task_index' %}?playbook_name={{ task.playbook.name }}">{{ task.playbook.name }}</a>
+    {% else %}
+      {{ task.playbook.name }}
+    {% endif %}
+  </td>
+  {% else %}
+  <td title="Search results for {{ task.playbook.path }}">
+    {% if not static_generation %}
+      <a href="{% url 'ui:task_index' %}?playbook_path={{ task.playbook.path }}">{{ task.playbook.path | truncatepath:25 }}</a>
```
Comment (dmsimard) [inline ara/ui/templates/partials/task_index_row.html:55]: Thinking out loud: I wonder whether this is the right link for this.

tl;dr: change from ``?playbook_path=playbook.yml`` to ``?playbook_path=playbook.yml&name=install%20apache`` ?

Right now clicking on a playbook's name (or path) would return every tasks for every playbook that matches the playbook name (or path).

Instead, it could provide a search for that task's name in addition to the playbook name (or path) such that it only provided this tasks' results across every playbook that match the playbook name (or path). There could be false positives due to the icontains nature of the search but it should otherwise be fairly precise.

File: ara/ui/templates/partials/task_index_row.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    </a>
+  </td>
+  <td>
+    <a {% if not static_generation %}href="{% url 'ui:task_index' %}?name={{ task.name }}" title="Search tasks for {{ task.name }}"{% endif %}>
+      {{ task.name }}
+    </a>
+  </td>
+  <td class="text-nowrap" title="{{ task.path }}">
+    {% if not static_generation %}
+      <a href="{% url 'ui:task_index' %}?path={{ task.path }}">{{ task.path | truncatepath:25 }}</a>
```
Comment (dmsimard) [inline ara/ui/templates/partials/task_index_row.html:37]: Same thought as playbook_path: https://github.com/ansible-community/ara/pull/342/files#r967882221, maybe we should include the task name when searching for the path so it doesn't return all the tasks in that file ?

**Comment by hille721** [comment]
> Right now the id is used to provide a direct link to tasks like this:

but for this we don't need to have the the field in the search box, this is working independently with a query parameter. 

Everything which is defined in `filters.py` can be used as query parameter, without a need to specify it on UI level in the search box. E.g.: https://demo.recordsansible.org/tasks?play=68994

**Review by hille721 — COMMENTED** [review]
I've added some comments with my findings.

Honestly I don't see many uses cases for this task page, so it is hard for me to give a valuable feedback. But I would guess the tasks pakge is there to compare task over different runs of a playbook and also over different playbooks.

But if you want to look in to the tasks of one single playbook run, imo you should directly go with the playbook report. That's why I don't see the need for a playbook id filter.

File: ara/ui/templates/host_index.html [hunk 4]
```diff
@@ -114,7 +120,7 @@
           <tr style="height:50px;">
             <th class="text-center" title="Status of the playbook">Playbook</th>
             <th class="text-center" title="Link to the host report">Report</th>
-            <th title="Last update">Last update {% include "partials/sort_by_date.html" with arg="updated" %}</th>
+            <th title="Date at which the host was first recorded">Date {% include "partials/sort_by_date.html" with arg="created" %}</th>
```
[the remark this answers is not in the document]
Comment (dmsimard) [reply ara/ui/templates/host_index.html:123]: I see what you mean and I wonder what we should put here.

I came to the realization that relying on the "updated" field is error prone since it gets updated whenever the row changes, including SQL migrations.

How would you describe the created date in a way that is appropriate for both hosts and latesthosts ?

File: ara/ui/templates/index.html [hunk 5]
```diff
@@ -134,8 +134,20 @@
...
+            <td class="text-center">
+              {% if not static_generation %}
+                <a href="{% url 'ui:host_index' %}?playbook={{ playbook.id }}" title="Search for hosts in this playbook">{{ playbook.items.hosts }}</a>
+              {% else %}
+                {{ playbook.items.hosts }}
+              {% endif %}
+            </td>
+            <td class="text-center">
+              {% if not static_generation %}
+              <a href="{% url 'ui:task_index' %}?playbook={{ playbook.id }}" title="Search for tasks in this playbook">{{ playbook.items.tasks }}</a>
```
[the remark this answers is not in the document]
Comment (dmsimard) [reply ara/ui/templates/index.html:146]: It's a different view: in the playbook report we actually have results (which are associated to tasks), not tasks themselves.

File: ara/ui/templates/index.html [hunk 6]
```diff
@@ -134,8 +134,20 @@
             </td>
             <td class="text-nowrap">{{ playbook.started | format_date }}</td>
             <td class="text-nowrap">{{ playbook.duration | format_duration }}</td>
-            <td class="text-center">{{ playbook.items.hosts }}</td>
-            <td class="text-center">{{ playbook.items.tasks }}</td>
+            <td class="text-center">
+              {% if not static_generation %}
+                <a href="{% url 'ui:host_index' %}?playbook={{ playbook.id }}" title="Search for hosts in this playbook">{{ playbook.items.hosts }}</a>
```
[the remark this answers is not in the document]
Comment (dmsimard) [reply ara/ui/templates/index.html:139]: Good catch on the missing ``latest=false``, thanks !

I'm not entirely convinced of the worth of this particular link (versus tasks which has value IMO) but it's "free" and I suppose since I was adding a link for tasks, I might as well add a link for hosts too.

I figure that it's not necessarily because I don't have a use case for this specific link that it doesn't mean it can't be useful to someone else, though.

**Comment by dmsimard** [comment]
> I've added some comments with my findings.
> 
> Honestly I don't see many uses cases for this task page, so it is hard for me to give a valuable feedback. But I would guess the tasks pakge is there to compare task over different runs of a playbook and also over different playbooks.
> 
> But if you want to look in to the tasks of one single playbook run, imo you should directly go with the playbook report. That's why I don't see the need for a playbook id filter.

Thanks for the review @hille721, much appreciated <3

Like you said, I believe the main use case for the tasks page is to search for the results of a particular task across multiple playbooks just like you might be interested in the results of a particular host across different playbooks.
That is sufficient justification on its' own for the page to exist, it isn't meant to replace the detailed results view from the playbook report :)

I still think there is value in the ability to filter tasks for a specific playbook ID but we've agreed that it's not really meant for humans. I'll reflect on that a bit.

File: ara/ui/templates/host_index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (hille721) [reply ara/ui/templates/host_index.html:123]: no idea. I checked, on the host index improvements I just made "Date" out of it :D 
image

File: ara/ui/templates/index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (hille721) [reply ara/ui/templates/index.html:144]: yes it is "free" but on the other side it increases complexity. Complexity for the users to navigate on the side (imo it can be confusing if there are many of links on one page), but also more complex for development. 
Afaik we don't have any tests which checks if links are still working in the way they are assumed to. Thus there is always the dangerour that links will be broken after ui updates because the developer didn't see that there also links pointing to a certain functionality, especially if the links are on other pages.

**Comment by hille721** [comment]
random thing what was comming into my mind yesterday:

we are having the playbooks index and `/playbooks/<playbook_id>.html`. We are also having the hosts index and `/hosts/<host_id>.html`.  Now we are building a tasks index page, but we don't have `/tasks/<task_id>.html`. 

Not sure what should be visivle on such a single task page, but could be maybe usefull.

But with the detailed `playbook_id` and `host_id` pages we are also having the possibility to link to them instead of filtering for them. E.g. instead of filtering the hosts index for a `playbook_id` directly go to the `/playbooks/<playbook_id>.html#hosts`. 

Further we are also having `/playbooks/<playbook_id>.html#results` and `/hosts/<host_id>.html#results`. But unforunately to `#tasks` section on those sites, as you already mentioned. But for one single host, tasks = results. 
Thus missing is just a tasks section on the playbook report page. Then we could also link to `/playbooks/<playbook_id>.html#tasks` instead of filtering the tasks for one playbook id.

File: ara/ui/templates/index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (hille721) [reply ara/ui/templates/index.html:151]: yeah I know, that's why I was writen about "task results" and "tasks" ;)
I've added one more general comment regarding those linking things: https://github.com/ansible-community/ara/pull/342#issuecomment-1245175356

**Comment by dmsimard** [comment]
> random thing what was comming into my mind yesterday:
> 
> we are having the playbooks index and `/playbooks/<playbook_id>.html`. We are also having the hosts index and `/hosts/<host_id>.html`. Now we are building a tasks index page, but we don't have `/tasks/<task_id>.html`.
> 
> Not sure what should be visible on such a single task page, but could be maybe usefull.

If it can be useful and there is a use case for it, I don't mind :p

The way that I look at it is that we have slowly been building UI views that are analog to API endpoints like:
- ``/api/v1/playbooks`` ­-> ``/`` (though we may argue one day whether it should live at ``/playbooks``)
- ``/api/v1/playbooks/<id>`` -> ``/playbooks/<id>.html``
- ``/api/v1/hosts`` (or ``/api/v1/latesthosts``) -> ``/hosts``
- ``/api/v1/hosts/<id>`` -> ``/hosts/<id>.html``
- etc

The CLI is the same with commands like
- ``ara playbook list`` -> ``/api/v1/playbooks``
- ``ara task list`` -> ``/api/v1/tasks``
- etc

> But with the detailed `playbook_id` and `host_id` pages we are also having the possibility to link to them instead of filtering for them. E.g. instead of filtering the hosts index for a `playbook_id` directly go to the `/playbooks/<playbook_id>.html#hosts`.
> 
> Further we are also having `/playbooks/<playbook_id>.html#results` and `/hosts/<host_id>.html#results`. But unforunately no `#tasks` section on those sites, as you already mentioned. But for one single host, tasks = results. Thus missing is just a tasks section on the playbook report page. Then we could also link to `/playbooks/<playbook_id>.html#tasks` instead of filtering the tasks for one playbook id.
> 
> Edit: I guess, this comment will be kinf of confusing. Just some random thoughts I wanted to share :)

I don't know if there is a need for an individual tasks block in the playbook report page.
I feel that the view calls for tasks to be "merged" with the hosts and results table to provide value but it does make it "special" in that regard.

It's possible I can't visualize what you mean, though, if you're convinced it could be good, maybe try a quick/dirty mockup.

File: ara/ui/templates/host_index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmsimard) [reply ara/ui/templates/host_index.html]: I updated it to ``Date this host was recorded`` which is not wrong and probably less confusing.

File: ara/ui/templates/index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmsimard) [reply ara/ui/templates/index.html:151]: I see what you mean.

The UI has grown organically, one page at a time, but it may eventually be time take a step back and consider the UX now that we have a lot more than just the playbook list and report page.

File: ara/ui/templates/partials/task_index_row.html (hunk 2 above)
Comment (dmsimard) [reply ara/ui/templates/partials/task_index_row.html:55]: I thought about it and let's go with the current implementation.
We can always make the search more precise later if we change our mind.

File: ara/ui/templates/partials/task_index_row.html (hunk 3 above)
Comment (dmsimard) [reply ara/ui/templates/partials/task_index_row.html]: I thought about it and let's go with the current implementation.
We can always make the search more precise later if we change our mind.

**Comment by dmsimard** [comment]
I've rebased the branch and updated the demo: https://demo.recordsansible.org/

I will let this simmer a bit (there is no hurry to merge it now) but I think it's good to go.
If we find other improvements or tweaks we can do them afterwards.

This PR has lived long enough, since Dec 4, 2021 :p

**Comment by dmsimard** [comment]
I had missed the static generation portion of it but it works well now. Let's merge it !
